Add expiration to banners

albertchae opened the pull request:
### What github issue is this PR for, if any?
Resolves #5604

### What changed, and _why_?

Add an optional expiration time to banners.
The timezone will be set to the creator's current timezone.
If a banner is active, but the time is past the expires_at, we will hide the banner.
If the banner is not active, we hide the banner regardless (existing behavior).
Also added displaying the banner on the banners index page

### How is this **tested**? (please write tests!) 💖💪
_Note: if you see a flake in your test build in github actions, please post in slack #casa "Flaky test: <link to failed build>" :) 💪_
_Note: We love [capybara](https://rubydoc.info/github/teamcapybara/capybara) tests! If you are writing both haml/js and ruby, please try to test your work with tests at every level including system tests like https://github.com/rubyforgood/casa/tree/main/spec/system_ 

Added tests for showing/hiding banner based on `expires_at`. Also added tests for setting `expires_at` in the banner new/edit form.

### Screenshots please :)
_Run your local server and take a screenshot of your work! Try to include the URL of the page as well as the contents of the page._ 
<img width="1728" alt="Screenshot 2024-05-31 at 1 50 04 PM" src="https://github.com/rubyforgood/casa/assets/217050/759c4f5d-4af5-4ef2-9a00-4121f4204a16">
<img width="1721" alt="Screenshot 2024-05-31 at 1 50 22 PM" src="https://github.com/rubyforgood/casa/assets/217050/96d4215e-c663-42eb-a38c-5c753ecdf12d">
<img width="1718" alt="Screenshot 2024-05-31 at 1 50 42 PM" src="https://github.com/rubyforgood/casa/assets/217050/4c4af923-0e0b-4b7f-96a0-a447d2700dec">


### Feelings gif (optional)
_What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:_
![clocks](https://media.giphy.com/media/Kvjugjvc1L9iE/giphy.gif)

albertchae committed:
Add expires_at to banners

albertchae committed:
Add expires_at to banners

albertchae committed:
Allow setting expires_at in create/update

We are choosing to set the timezone in the user's timezone (defined by
`cookies[:browser_time_zone]`) by default

albertchae committed:
Use expires_at to check whether a banner should be displayed or not

albertchae committed:
add test for showing/hiding banner

albertchae committed:
Add system test for expiration time

albertchae committed:
Display expires at on index page

albertchae committed:
Convert time stored in UTC to local timezone for picker

form.datetime_field uses html input `datetime-local` under the hood, which does not factor in
timezones, so convert it here explicitly

elasticspoon commented on spec/requests/banners_spec.rb:
Do you think it makes sense to add a test verify the time zone stuff? Something along the lines of: 
Set banner => visit page, it shows => change cookies timezone => visit again and confirm that is it gone?

albertchae replied:
>  Set banner => visit page, it shows => change cookies timezone => visit again and confirm that is it gone?

Hmm I was thinking of doing something similar but I don't think the scenario you describe would happen since the expiration time is a fixed point in time, so changing timezone would only affect how it gets displayed. So I did try to test this scenario

`Timezone in PT, set banner to N => visit banner edit and see N:00 => change cookies timezone to ET => visit banner edit and see N+3`

But was not sure how to do this in capybara/selenium. This is what I tried without success

```ruby
it "displays expiration time in your current time zone" do
  Capybara.using_session("America/Los_Angeles") do
    sign_in admin

    visit banners_path
    click_on "New Banner"
    fill_in "Name", with: "name"
    fill_in "banner_expires_at", with: "06082024\t0710pm"
    click_on "Submit"

    visit banners_path
    expect(page).to have_text("name")

    visit banners_path
    within "#banners" do
      click_on "Edit", match: :first
    end
    expect(page).to have_field("banner_expires_at", with: "2024-06-08T19:10")
  end

  Capybara.using_session("America/New_York") do
    sign_in admin

    visit banners_path
    within "#banners" do
      click_on "Edit", match: :first
    end
    expect(page).to have_field("banner_expires_at", with: "2024-06-08T22:10")
  end
end
```

elasticspoon reviewed (approved):
LGTM. Not sure how you feel about potentially adding a test to validate cookie time zone behavior.

Also a note for future me

![image](https://github.com/rubyforgood/casa/assets/14540596/0fe3128a-e8ab-4a40-968c-7ae78f4b1069)

it may make more sense to have a checkbox for the expiration that dictates whether than field should be shown at all.

schoork commented on app/controllers/application_controller.rb:
```diff
@@ -37,7 +37,9 @@ def set_active_banner
 
     @active_banner = current_organization.banners.active.first
+
     @active_banner = nil if session[:dismissed_banner] == @active_banner&.id
+    @active_banner = nil if @active_banner&.expires_at && Time.now.in_time_zone(cookies[:browser_time_zone]) > @active_banner&.expires_at
```
Let's extract this to a single method and then test against it.

albertchae replied:
extracted it to `Banner#expired` and added a test. Also switched to `Time.current` per https://thoughtbot.com/blog/its-about-time-zones.

I also realized that we don't need the timezone conversion here because `Time.current`/`Time.now` are computed in the backend and `>` is timezone aware, so I removed that as well.

schoork commented on app/controllers/banners_controller.rb:
```diff
@@ -65,7 +65,12 @@ def set_banner
 
   def banner_params
-    params.require(:banner).permit(:active, :content, :name).merge(user: current_user)
+    params.require(:banner).permit(:active, :content, :name, :expires_at).merge(user: current_user)
+      .tap { |banner_params| set_expires_at_in_user_time_zone(banner_params) }
```
This is doing two things. Can we split the call to conversion out of the `banner_params` method? You might take a look at the `CaseContactParameters` class.

albertchae replied:
Moved this out to a `BannerParameters` class

schoork commented on app/controllers/banners_controller.rb:
```diff
@@ -65,7 +65,12 @@ def set_banner
+  end
+
+  def set_expires_at_in_user_time_zone(banner_params)
+    banner_params[:expires_at] = banner_params[:expires_at].in_time_zone(cookies[:browser_time_zone])
```
Is this method testable? Can we make it testable and then write some tests for it? Maybe this should be a method on the model.

albertchae replied:
I believe the tests for `BannerParameters` cover this now. I debated putting it on the model originally but formatting parameters from a request seems like a controller responsibility, so I think having it in the controller/params object makes sense but let me know if you think otherwise

schoork commented on app/views/banners/_form.html.erb:
```diff
@@ -26,6 +26,10 @@
     </div>
+    <span class="input-style-1">
+      <%= form.label :expires_at, "Expires at (optional)" %>
+      <%= form.datetime_field :expires_at, value: banner.expires_at&.in_time_zone(cookies[:browser_time_zone]), required: false %>
```
This logic should be in a view helper (and tested).
```
banner.expires_at&.in_time_zone(cookies[:browser_time_zone])
```
You might take a look at [datetime_local_field](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-datetime_local_field). It might eliminate the need for the logic.

albertchae replied:
Unfortunately I don't think datetime_local_field helps us here.  According to your link, it wraps [ActionView::Helpers::FormHelper#datetime_local_field](https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-datetime_local_field) which in turn states that it's only an alias for `datetime_field`. According to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#setting_timezones, timezone aware logic was intentionally left out of this element.

That said, I did try to move this logic to `BannerHelper` but ultimately put it in the model. It kind of feels like it would fit best in a presenter type object but might be overkill for now? Any thoughts?

schoork replied:
Either model or presenter is fine. I agree presenter makes more sense, but if it's just one method does feel like overkill.

schoork commented on app/views/banners/index.html.erb:
```diff
@@ -43,6 +44,13 @@
                   <% end %>
                 </td>
+                <td class="min-width">
+                  <% if banner.expires_at %>
+                    <%= distance_of_time_in_words(Time.now, banner.expires_at) %>
+                  <% else %>
+                    No
+                  <% end %>
```
A great candidate for a method in the model or view helper (with tests).

albertchae replied:
Moved this to `BannerHelper` which also helped me realize I was missing the case for when the expiration had passed, so I fixed that too

schoork commented on spec/requests/banners_spec.rb:
```diff
@@ -47,4 +47,28 @@
+      end
+    end
+
+    context "when expires_at is after today" do
```
Minor: I think you mean **before** here.

albertchae replied:
fixed

albertchae committed:
Factor out expired? method onto Banner model

albertchae committed:
Factor out parameter handling for Banners to BannerParameters

albertchae committed:
Move timezone conversion for frontend display to model method

albertchae committed:
Helper method for when a banner expires for index page

albertchae committed:
Add clarifying comments on "with" or "in" time_zone

Frontend->Backend: We don't have timezone so we have to attach timezone from browser
Backend->Frontend: We have timezone UTC and need to convert it to the
browser timezone before displaying

compwron reviewed (approved):
nice
